Adopt code for Khalani Protocol

serejke opened the pull request:
This PR contains the patches on top of the origin Avalanche Faucet that adopt for Khalani protocol:
- allow to configure faucet from a file specified by `CONFIG_FILE`
- allow to override RPC URLs with ENV variables of name `EVM_CHAINS_${ID}_RPC`, where `ID` is the chain ID specified in `evmchains` of `config.json`, for example `EVM_CHAINS_KHALANITESTNET_RPC`
- allow to disable captcha verification — needed to test the deployment and then register Google Captcha for the production domain
- added Khalani Protocol branding (logo / docs)
- GitHub pipeline to build a Docker image

Closes https://github.com/tvl-labs/infrastructure/issues/213
Closes https://github.com/tvl-labs/infrastructure/issues/214

serejke committed:
chore: add .dockerignore

serejke committed:
fix: non-null field

serejke committed:
chore [chain]: initial "khalanitestnet" config.

serejke committed:
chore: cleanup .github

serejke committed:
chore: Khalani branding.

serejke committed:
chore: update README

serejke committed:
chore: allow to disable captcha verification.

serejke committed:
chore: read config from a file.

serejke committed:
chore: delete unused field

serejke committed:
chore: provide RPC URLs from env variables.

serejke committed:
chore: do not create a captcha object when captcha is disabled.

serejke committed:
docs: update README

serejke committed:
chore: add Docker pipeline and CODEOWNERS.

serejke committed:
chore: update .env.dist example

serejke committed:
docs: update README.md

distributedstatemachine commented on .env.dist:
```diff
@@ -1,3 +1,5 @@
-CAPTCHA_SECRET="Google ReCaptcha V3 Secret"
-V2_CAPTCHA_SECRET="Google ReCaptcha V2 Secret"
+EVM_CHAINS_KHALANITESTNET_RPC=https://www.axon-node.info
+DISABLE_CAPTCHA=true
```
Why are you disabling captcha ? Faucets get DDOSed (tbh its a nice problem for us to have at this point ) . I would like to understand the intuition behind this

distributedstatemachine replied:
seen this : 

> - allow to disable captcha verification — needed to test the deployment and then register Google Captcha for the production domain

distributedstatemachine commented on client/public/index.html:
```diff
@@ -2,21 +2,22 @@
-    <meta property="og:url" content="https://faucet.avax.network/">
-    <meta property="og:description" content="Get testnet tokens to use on the Avalanche Fuji Network (testnet) or a Subnet (WAGMI, Defi Kingdoms, Dexalot, Swimmer, and Castle Crush) in one place.">
+
+    <meta name="description" content="Get testnet tokens to use for the Khalani Protocol (USDC/USDT testnet tokens on Khalani chain, Goerli, Avalanche Fuji, etc) in one place.">
```
Do you want to take out chain specific details  `USDC/USDT testnet tokens on Khalani chain, Goerli, Avalanche Fuji, etc` and replace it with something along the lines of "Testnet tokens to interact with the Khalani protocol"

distributedstatemachine commented on client/public/index.html:
```diff
@@ -2,21 +2,22 @@
+    <meta property="og:site_name" content="Faucet | Khalani">
+    <!-- TODO: set the site URL -->
+    <meta property="og:url" content="https://khalani-faucet.digipnyx.org/">
+    <meta property="og:description" content="Get testnet tokens to use for the Khalani Protocol (USDC/USDT testnet tokens on Khalani chain, Goerli, Avalanche Fuji, etc) in one place.">
```
Same as above

serejke committed:
fix: read .env file first

serejke committed:
docs: development readme

distributedstatemachine reviewed (changes requested):
This need to be modified to `Add chain to Metamask`, as subnets are an avalanche specific thing

serejke committed:
cleanup: drop unused TOTP.

serejke committed:
refactor: decouple configuration into functions.

serejke committed:
fix: make sure the EVM is initialized.

serejke committed:
refactor: simplify if conditions.

serejke committed:
refactor: extract variable.

serejke committed:
refactor: specify type.

serejke committed:
refactor: make the function return <T>

serejke committed:
refactor: add types.

serejke committed:
refactor: add type.

serejke committed:
refactor: add type.

serejke committed:
refactor: inline function.

serejke committed:
refactor: add types.

serejke committed:
refactor: add types.

serejke committed:
fix: wait for all balances to be updated.

serejke committed:
cleanup: move log to a more appropriate place.

serejke committed:
fix: simplify recalibration states.

serejke committed:
refactor: lowercase file name

serejke committed:
refactor: return type.

serejke committed:
refactor: identify errors by request ID.

serejke committed:
refactor: reuse type definition.

serejke committed:
fix: streamline state management of requests.

Use single Map to store all possible states of a request.

serejke committed:
refactor: replace callback with Promise.

serejke committed:
refactor: rename variable

serejke committed:
refactor: use config for static parameters

serejke committed:
logs: wording

serejke committed:
refactor: count requests by the map size

serejke committed:
refactor: flip if-condition

serejke committed:
refactor: rename and rethrow error.

serejke committed:
refactor: rename function

serejke committed:
refactor: pull from wait queue during recalibration.

serejke committed:
refactor: pull from the status map.

serejke committed:
refactor: rename

serejke committed:
refactor: reorder fields

serejke committed:
refactor: simplify error handling

serejke committed:
logs: current balances.

serejke committed:
fix: workaround frontend bug sending ERC20 parameter for native token.

serejke committed:
fix: streamline the state management.

serejke committed:
refactor: decouple signer interface.

serejke committed:
feat: support AWS KMS signer.

serejke committed:
logs: JSON logs and timestamps.

serejke committed:
feat: add Prometheus metrics.

serejke committed:
refactor: use global simple request ID.

serejke committed:
docs: update README

serejke committed:
docs: update README

serejke committed:
docs: wording

serejke committed:
docs: use "chain" instead of "subnet"

serejke commented:
A lot of code has been re-implemented on the server side:
- properly handle errors
- significantly streamlined the state management of requests and nonce/balance recalibration (`mem-pool` -> `signed` -> `sent` OR `error`)
- fixed a critical blocker that the code used to sign the transaction twice
- added support of faucet wallet signer with [AWS KMS](https://docs.aws.amazon.com/kms/latest/APIReference/API_Sign.html)
- added Prometheus support 

Next steps:
- deploy this PR with full KMS and Prometheus EKS setup
- see if captcha / rate limit changes are necessary, if not then it can be merged

konraddb commented on server.ts:
```diff
@@ -1,184 +1,225 @@
 
-    let balance: BN = evm?.instance.getBalance(erc20)
+    // GET request for fetching faucet address for the specified chain
+    router.get('/faucetAddress', async (req: any, res: any) => {
```
no need 'async' here

konraddb commented on server.ts:
```diff
@@ -1,184 +1,225 @@
 
-    res.status(200).send({
-        balance: balance?.toString()
+        const balance = evm.getBalance(id);
```
In the source code, when there is no balance, their are using `balance = new BN(0)`

konraddb commented on client/public/banner.jpeg:
Can we perhaps use a logo here with a transparent background and of better quality?

konraddb reviewed (commented):
I compared the source code with the refactored code and did not find any potential issues. However, by the way, a few cosmetic issues have been reported.

serejke committed:
logs: add chain name.

serejke committed:
cleanup: unused image.

serejke committed:
refactor: type files

serejke committed:
refactor: cleanup dependencies and imports

serejke committed:
logs: improve logs for failures

serejke committed:
fix: simplify nonce management.

serejke committed:
feat: log headers.

serejke committed:
fix: support IP rate limiting in Kubernetes.

serejke committed:
fix: update captcha key

serejke committed:
fix: workaround for unhandled RPC errors.

https://github.com/odanado/cloud-cryptographic-wallet/issues/845

serejke committed:
fix: CR comment.

serejke committed:
fix: update logo quality.

serejke committed:
cleanup: unused "usage" endpoint.

serejke committed:
cleanup: client README.

serejke committed:
chore: preconfigure development wallet.

serejke committed:
fix: update GitHub link

serejke committed:
refactor: simplify code.

serejke committed:
fix: do not abort cancel unnecessarily.

distributedstatemachine commented:
@serejke is this good to go?

serejke commented:
Merged 🚀
